### ISSUE TITLE
Load data from gift-data repos

### ISSUE THREAD

**Issue by risenW** [datopian/gift-portal#8]
Load data from a specified repository, parse it and cache it in-app metastore displaying in Gift-portal pages

## Acceptance

Given a list of datasets specified in a simple text file or JSON as the configuration in the application e.g.

```
https://github.com/gift-data/presupuesto-mexico-2008-2019
https://github.com/gift-data/croatia-budget-spending
```

* [x] Those datasets show up on the site where we would expect __Done__ https://gift-portal-git-feature-metastore.datopian1.vercel.app/
* [x] Errors in loading those datasets (or parsing their metadata) are displayed on site prominently (e.g. banner on home page ...?) __Done__ The third dataset shown in the [live site](https://gift-portal-git-feature-metastore.datopian1.vercel.app/) has a parsing error and is displayed in the card

In more detail:

* [x] The dataset with its metadata is loaded from github into the "local Metastore" on application startup  __Done__ https://github.com/datopian/gift-portal/pull/25
* [ ] Resource data that is stored into LFS storage has its urls dereferenced in an appropriate way so that they can be used in rendering the site (TODO: probably needs clarification - This is a separate issue: see similar issue #22
* [ ] (BONUS) there is a "api" url e.g. /admin/reload which will reload the DB (this means we can reload the datasets without having to reboot the app)

## Tasks

* [x] Add config file (0.0.5 day)
~~* [ ] Init catalog (once on app load) (0.05 day)~~ This is handled in #18 
  * [x] Load data from specified repos using graphql (0.2 day)
  * [x] Parse JSON of individual datasets (each repo datapackage.json) (0.3 day)
  * [x] Add the README onto the data package object on `readme` attribute (0.1 day)
* [x]  Update home and dataset page (this should not really be necessary as done in  #18) (0.1 day)
* [ ] Add API url that re-inits the metastore (0.2 day) This depends on #18 
 
## Analysis

* Add lists of repositories names to load data from in a config file:
* 
```json
{
  "1":  "presupuesto-mexico-2008-2019",
  "2": "croatia-budget-spending"
}  
```

* Load json config file and for each repository name, retrieve its content including datapackage.json:
    * We can use the [Github graphql api](https://docs.github.com/en/free-pro-team@latest/graphql) for loading data from repository. The query below was tested in the [Github graphql explorer](https://docs.github.com/en/free-pro-team@latest/graphql/overview/explorer), and it will return all the contents in a repository including metadata:
```graphql
    query RepoFiles($owner: String!, $name: String!) {
      repository(owner: $owner, name: $name) {
        description
        url
        createdAt
        description
        updatedAt
        resourcePath
        name
        object(expression: "master:") {
          ... on Tree {
            entries {
              name
              type
              object {
                ... on Blob {
                  byteSize
                  text
                }
              }
            }
          }
        }
      }
    }
```

Where query parameters is the repository owner     (gift-data), and the repository name (presupuesto-    mexico-2008-2019). Below a sample response from the query above:

```json
{
  "data": {
    "repository": {
        "description": null,
          "url": "https://github.com/gift-data/presupuesto-de-uruguay",
          "createdAt": "2020-08-31T16:30:44Z",
          "homepageUrl": null,
          "updatedAt": "2020-09-01T18:14:59Z",
          "resourcePath": "/gift-data/presup",
          "object": {
            "entries": [
              {
                "name": ".gitattributes",
                "type": "blob",
                "object": {
                  "byteSize": 42,
                  "text": "*.csv filter=lfs diff=lfs merge=lfs -text\n"
                }
              },
              {
                "name": ".lfsconfig",
                "type": "blob",
                "object": {
                  "byteSize": 75,
                  "text": "[lfs]\n\turl = http://127.0.0.1:8080/gift-data/presupuesto-mexico-2008-2019/\n"
                }
              },
              {
                "name": "data",
                "type": "tree",
                "object": {}
              },
              {
                "name": "datapackage.json",
                "type": "blob",
                "object": {
                  "byteSize": 196968,
                  "text": "{\n  \"@context\": \"http://schemas.frictionlessdata.io/fiscal-data-package.jsonld\",\n  \"bytes\": 7226941044,\n  \"count_of_rows\": 11951209,\n  \"hash\": \"1e5a26d6be2b3068fa1d0d3a44a6d3e6\",\n  \"model\": {\n    \"dimensions\": {\n      \"activity\": {\n        \"attributes\": {\n          \"DESC_MODALIDAD\": {\n            \"labelfor\": \"ID_MODALIDAD\",\n            \"source\": \"DESC_MODALIDAD\",\n            \"title\": \"DESC_MODALIDAD\"\n          },\n          \"DESC_PP\": {\n            \"labelfor\": \"ID_PP\",\n            \"source\": \"DESC_PP\",\n            \"title\": \"DESC_PP\"\n          },\n          \"ID_MODALIDAD\": {\n            \"source\": \"ID_MODALIDAD\",\n            \"title\": \"Modalidad\"\n          },\n          \"ID_PP\": {\n            \"parent\": \"ID_MODALIDAD\",\n            \"source\": \"ID_PP\",\n            \"title\": \"Programa Presupuestario\"\n          }\n        },\n        \"dimensionType\": \"activity\",\n        \"primaryKey\": [\n          \"ID_MODALIDAD\",\n          \"ID_PP\"\n        ]\n      },\n      \"administrative-classification\": {\n        \"attributes\": {\n          \"DESC_RAMO\": {\n            \"labelfor\": \"ID_RAMO\",\n            \"source\": \"DESC_RAMO\",\n            \"title\": \"DESC_RAMO\"\n          },\n          \"DESC_UR\": {\n            \"labelfor\": \"ID_UR\",\n            \"source\": \"DESC_UR\",\n            \"title\": \"DESC_UR\"\n          },\n          \"ID_RAMO\": {\n            \"source\": \"ID_RAMO\",\n            \"title\": \"Ramo\"\n
                  
}
```
* For each Graphql result returned:
    * parse the datapackage.json object using (JSON.parse)
    * retreive metadata (file url, title, name etc)
    * Construct resource urls for from Git-lfs files (Analysis below)
    * Get small sample with frictionless.js using constructed url link
    * Save in local db with the repo name (Analysis for local db is done below). Example of db format can be:

```json=
{
    "presupuesto-mexico-2008-2019": {
        createdAt: "2020-08-31T16:30:44Z",
        updatedAt: "2020-09-01T18:14:59Z""
        url:"https://github.com/gift-data/presupuesto-de-uruguay"
        sample: [[...], [...]]
        resource_urls: ["https://some.storage.gift-data/presupuesto-de-uruguay1.csv",
                        "https://some.storage.gift-data/presupuesto-de-uruguay2.csv"]
        title: "presupuesto-mexico-2008-2019"
        name: "presupuesto-mexico-2008-2019"
        description: "This is a sample test file"
    }
}

```
The fields above are enough for what we are currently displaying in the datasets and home page of gift portal

**Comment by sglavoie** [comment]
@rufuspollock, I wanted to flag that @risenW told me today during our daily stand-up he would need your feedback on this issue. :+1:

**Comment by rufuspollock** [comment]
@risenW i need more analysis here including acceptance etc. The main thing here is we need to actually retrieve the *contents* of files in the repo - is that possible with the graph API? (Specifically we need to a) retrieve contents of a repo b) retrieve contents of a file in the repo e.g. `datapackage.json`)

**Comment by risenW** [comment]
> @risenW i need more analysis here e.g. more detailed acceptance (i have done some updating for you). The main thing here is we need to actually retrieve the _contents_ of files in the repo - is that possible with the graph API? (Specifically we need to a) retrieve contents of a repo b) retrieve contents of a file in the repo e.g. `datapackage.json`)

Okay, following your acceptance criteria, I'll provide more analysis. As regards the query result from Graphql, Yes, we get back the repos as well as their files, including the datapackage.json

**Comment by rufuspollock** [comment]
@risenW 

> Okay, following your acceptance criteria, I'll provide more analysis. As regards the query result from Graphql, Yes, we get back the repos as well as their files, including the datapackage.json.

Great. Can you detail the graphql query with link to github api docs in the analysis section 

> Also, we do not need to specify the Github URL links, we can retrieve them all when we make a call using the organization ID

I don't want to do that - i want an explicit list of github repos (e.g. in case we use use repos from multiple orgs, or we want to exclude a repo from an org etc ...)

**Comment by risenW** [comment]
> Great. Can you detail the graphql query with link to github api docs in the analysis section
> 
> >
Yea, working on that now



> I don't want to do that - i want an explicit list of github repos (e.g. in case we use use repos from multiple orgs, or we want to exclude a repo from an org etc ...)

Okay, we can do this as well with graphql. I'll update the steps in the analysis as well

**Comment by risenW** [comment]
@rufuspollock I have updated the analysis for this issue

**Comment by sglavoie** [comment]
FIXED: The original set of tasks has been completed.

**Comment by rufuspollock** [comment]
@risenW @steveoni why do we have `/db/` and specifically `db/db.js` as well as `lib/metastore.js` - how are these two related?
